src/gadgets/r1cs.rs

```diff
@@ -338,6 +338,101 @@ impl<G: Group> AllocatedRelaxedR1CSInstance<G> {
     })
   }
 
+ /// Folds self with a relaxed r1cs instance and returns the result
+ #[allow(clippy::too_many_arguments)]
+ #[allow(unused)]
+ pub fn fold_with_relaxed_r1cs<CS: ConstraintSystem<<G as Group>::Base>>(
+   &self,
+   mut cs: CS,
+   params: AllocatedNum<G::Base>, // hash of R1CSShape of F'
+   u: AllocatedRelaxedR1CSInstance<G>,
+   T: AllocatedPoint<G>,
+   ro_consts: ROConstantsCircuit<G>,
+   limb_width: usize,
+   n_limbs: usize,
+ ) -> Result<AllocatedRelaxedR1CSInstance<G>, SynthesisError> {
+   // Compute r:
+   let mut ro = G::ROCircuit::new(ro_consts, NUM_FE_FOR_RO);
+   ro.absorb(params);
+   self.absorb_in_ro(cs.namespace(|| "absorb running instance"), &mut ro)?;
+   u.absorb_in_ro(cs.namespace(|| "absorb running instance u"), &mut ro)?;
+   ro.absorb(T.x.clone());
+   ro.absorb(T.y.clone());
+   ro.absorb(T.is_infinity.clone());
+   let r_bits = ro.squeeze(cs.namespace(|| "r bits"), NUM_CHALLENGE_BITS)?;
+   let r = le_bits_to_num(cs.namespace(|| "r"), r_bits.clone())?;
+
+   // W_fold = self.W + r * u.W
+   let rW = u.W.scalar_mul(cs.namespace(|| "r * u.W"), r_bits.clone())?;
+   let W_fold = self.W.add(cs.namespace(|| "self.W + r * u.W"), &rW)?;
+
+   // E_fold = self.E + r * T + r * r * U.E
+   let rT = T.scalar_mul(cs.namespace(|| "r * T"), r_bits.clone())?;
+   let r_e_2 = u.E.scalar_mul(cs.namespace(|| "r * E_2"), r_bits.clone())?;
+   // Todo - there has to be a better way than 2 scalar mul
+   let r_squared_e_2 = r_e_2.scalar_mul(cs.namespace(|| "r * r * E_2"), r_bits)?;
+   let rT_plus_r_squared_E_2 = rT.add(cs.namespace(|| "r * r * E_2"), &r_squared_e_2)?;
+   let E_fold = self.E.add(cs.namespace(|| "self.E + r * T"), &rT_plus_r_squared_E_2)?;
+
+   // u_fold = u_r + r
+   let u_fold = AllocatedNum::alloc(cs.namespace(|| "u_fold"), || {
+     Ok(*self.u.get_value().get()? + r.get_value().get()?)
+   })?;
+   cs.enforce(
```

Comment:
`.enforce()` do check `A*B = C` to the right constraint should be:

```rust
   // u_fold = u_r + r
    cs.enforce(
      || "Check u_fold",
      |lc| lc + u_r*(1/r),
      |lc| lc + CS::one(),
      |lc| lc + u_fold.get_variable(),
    );
```

Reply:
```
cs.enforce(
  || "check i + 1",
  |lc| lc,
  |lc| lc,
  |lc| lc + i_new.get_variable() - CS::one() - i.get_variable(),
);
```
Found this in original code, it worked for `A = B = C = lc`.

Reply:
Some more detail https://github.com/filecoin-project/bellperson/blob/master/src/gadgets/boolean.rs#L49-L60